examples/object_subclass/author.rs

```diff
@@ -20,11 +19,17 @@ mod imp {
         surname: RefCell<String>,
     }
 
+    #[glib::signals(wrapper_type = super::Author)]
+    impl Author {
+
+        #[signal]
+        fn awarded(&self) {}
```

Comment:
I wonder if it would be possible to make

```suggestion
        fn awarded(&self);
```

and that means that the signal has no class handler, and

```suggestion
        fn awarded(&self) { println!("stuff"); }
```

which means that it has a class handler with the body of the function.

In the first case, the macro would have to remove the whole function after expansion.

Reply:
A function without a block isn't valid rust syntax outside of an `extern` block. I don't want to add some invalid Syntax that would be removed by the macro.

It's also way more complicated to parse this. I would have to write a custom parser based on `syn::ItemImpl` that accepts these "invalid" functions. And it confuses other parsing code for other macros or rustfmt.

I can add and parse a toplevel pseudo-macro, meaning something like this would be ok:

```rs
#[glib::signals(wrapper_type = super::FooObject)]
impl FooObject {
    signals! {
        fn awarded() -> i32;
    }
}
```

This pseudo-macro is valid Rust Syntax and should work with rustfmt and other macros.

Reply:
Maybe instead of using an `impl Object`, the usage could be `trait ObjectSignals` making it possible to provide a default impl?

Reply:
Default implementations can only use methods provided by the trait's required subtraits and itself.

The class handler implementations would have to reside in an `impl ObjectSignals for FooObject` block.

I don't really see how I'd map a trait to a definition of signals, as I don't want the macro to modify the trait in ways that make it hard for a developer to understand what the trait will actually look like.

The problem I see is that a semicolon instead of a block in a trait definition means that you **have to provide the class handler** in your implementation of that trait, while you want it to mean that **there is no class handler**.

Reply:
The trait itself wouldn't be used. You can then `impl T for SomeObject` and forward the calls when the trait has an impl, no? 

Anyways, zbus is using something similar for it zbus::proxy! macro.

Reply:
There's an open issue around [`#[zbus::proxy]` eating the trait definition](https://github.com/dbus2/zbus/issues/225).
I don't want to introduce these weird macro-related edge cases into `glib`.

What do you think about this?

```rs
#[glib::signals(wrapper_type = super::FooObject)]
trait FooObjectSignals {
    fn simple_signal(&self);
    
    fn complex_signal_without_class_handler(&self) -> i32;
    
    #[signal(class_handler, run_first]
    fn complex_signal_with_class_handler(&self, param: i32) -> i32;
}

impl FooObjectSignals for FooObject {
    fn complex_signal_with_class_handler(&self, param: i32) -> i32 {
        10
    }
}
```

I would make the macro produce something like:

```rs
trait FooObjectSignals: glib::subclass::types::ObjectSubclass<Type = super::FooObject> {
    // These functions have to be removed from the trait since they doesn't have a class handler one could implement:
    //     simple_signal(&self); 
    //     complex_signal_without_class_handler(&self) -> i32;
    // Providing a function body for signals without class handlers is forbidden. 
    
    // This function is allowed to have a default implementation but does not need to have one.    
    fn complex_signal_with_class_handler(&self) -> i32;
    
    #[doc(hidden)]
    fn __derived_signals() {
        static SIGNALS: std::sync::OnceLock<[glib::subclass::signal::Signal; 3]> = std::sync::OnceLock::new();
        SIGNALS.get_or_init(|| [
            glib::subclass::signal::Signal::builder("simple_signal").build(),
            glib::subclass::signal::Signal::builder("complex_signal_without_class_handler")
                .param_types([<i32 as glib::types::StaticType>::static_type()])
                .return_type::<i32>()
                .build(),
            glib::subclass::signal::Signal::builder("complex_signal_with_class_handler")
                .param_types([<i32 as glib::types::StaticType>::static_type()])
                .return_type::<i32>()
                .class_handler(|params| {
                    Some(<i32 as glib::value::ToValue>::to_value(
                        &Self::complex_signal_with_class_handler(
                            glib::subclass::types::ObjectSubclassExt::from_obj(&params[0].get::<Self::Type>().unwrap()),
                            params[1].get::<i32>().unwrap(),
                        )
                    )
                })
                .build(),
        ])
    }
}
```